UserProfile.js

```diff
@@ -0,0 +1,259 @@
+import * as L from "./Logger.js";
+import * as Validate from "./Validators.js";
+
+class UserProfile {
+    #state;
+
+    constructor(state) {
+        this.#state = state;
+    }
+
+    isValidUserProfileCall = function (keyName, keyValue, functionName) {
+        if (!this.#state.isInitialized) {
+            L.w(`${functionName}, 'init' must be called before ${functionName}`);
+            return false;
+        }
+        // validate keyName
+        if(!keyName) {
+            return false;
+        }
+        if (typeof keyName !== 'string' || keyName.trim() === '') {
+            L.w(`${functionName}, provided keyName is not a valid string`);
+            return false;
+        }
+        // validate keyValue
+        if (keyValue === null || !Validate.isValidPrimitiveOrArray(keyValue)) {
+            L.w(`${functionName}, provided keyValue is not valid`);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * 
+     * Set custom key and value pair for the current user
+     * 
+     * @param {string} keyName - user property key
+     * @param {object} keyValue - user property value
+     * @returns {void}
+     */
+    setProperty = async function (keyName, keyValue) {
+        if(!this.isValidUserProfileCall(keyName, keyValue, "setProperty")) {
+            return;
+        }
+        let formattedKeyValue = keyValue.toString();
+        L.d(`setProperty, Setting user property: [${keyName}, ${formattedKeyValue}]`);
+        await this.#state.CountlyReactNative.userData_setProperty([keyName, formattedKeyValue]);
+    };
+
+    /**
+     * 
+     * Set predefined and/or custom key and value pairs for the current user
+     * 
+     * @param {object} userData - custom key value pairs
+     * @returns {void}
+     */
+    setProperties = async function (userData) {
+        if (!this.#state.isInitialized) {
+            L.w("setProperties, 'init' must be called before 'setProperties'");
+            return;
+        }
+        if (!userData) {
+            L.w("setProperties, User profile data should not be null or undefined");
+            return;
+        }
+        if (typeof userData !== "object") {
+            L.w(`setProperties, unsupported data type of user data '${typeof userData}'`);
+            return;
+        }
+        L.d(`setProperties, Setting properties: [${JSON.stringify(userData)}]`);
+        const predefinedKeys = {
```

Comment:
predefined keys might be initialized at the top of the class?

Reply:
i don't think that's necessary it's only used in here now, in the future if required agreed on that

Reply:
that is correct but this will reinitailize the array wevery time the function is called.